[SPARK-23975][ML]Allow Clustering to take Arrays of Double as input features

lu-wang-dl opened the pull request:
## What changes were proposed in this pull request?

- Multiple possible input types is added in validateAndTransformSchema() and computeCost() while checking column type

- Add if statement in transform() to support array type as featuresCol

- Add the case statement in fit() while selecting columns from dataset

These changes will be applied to KMeans first, then to other clustering method

## How was this patch tested?

unit test is added

Please review http://spark.apache.org/contributing.html before opening a pull request.

lu-wang-dl committed:
add Array input support for KMeans

lu-wang-dl committed:
remove redundent code

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #89410 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89410/testReport)** for PR 21081 at commit [`badb0cc`](https://github.com/apache/spark/commit/badb0cc5ca6ca69bb8e8fc0fce5ea05a4100bca0).

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/clustering/KMeansSuite.scala:
```diff
@@ -194,6 +195,34 @@ class KMeansSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultR
+
+    assert(kmeans.getK === 2)
+    assert(kmeans.getFeaturesCol === featuresColName)
+    assert(kmeans.getPredictionCol === "prediction")
```
No need to check this or the other Params which are not relevant to this test

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/clustering/KMeansSuite.scala:
```diff
@@ -194,6 +195,34 @@ class KMeansSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultR
+    val arrayUDF = udf { (features: Vector) =>
+      features.toArray
+    }
+    val newdataset = dataset.withColumn(featuresColName, arrayUDF(col("features")) )
```
nit: You could drop the original column as well just to make extra sure that it's not being accidentally used.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/clustering/KMeansSuite.scala:
```diff
@@ -194,6 +195,34 @@ class KMeansSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultR
+    assert(kmeans.getDistanceMeasure === DistanceMeasure.EUCLIDEAN)
+    val model = kmeans.setMaxIter(1).fit(newdataset)
+
+    MLTestingUtils.checkCopyAndUids(kmeans, model)
```
You don't need this test here

jkbradley replied:
ditto for hasSummary and copying

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/clustering/KMeansSuite.scala:
```diff
@@ -194,6 +195,34 @@ class KMeansSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultR
     assert(e.getCause.getMessage.contains("Cosine distance is not defined"))
   }
 
+  test("KMean with Array input") {
```
This should check transform and computeCost too

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -123,8 +128,15 @@ class KMeansModel private[ml] (
-    val predictUDF = udf((vector: Vector) => predict(vector))
-    dataset.withColumn($(predictionCol), predictUDF(col($(featuresCol))))
+    // val predictUDF = udf((vector: Vector) => predict(vector))
+    if (dataset.schema($(featuresCol)).dataType.equals(new VectorUDT)) {
```
tip: This can be more succinct if written as:
```
val predictUDF = if (dataset.schema(...).dataType.equals(...)) { A } else { B }
dataset.withColumn($(predictionCol), predictUDF(col($(featuresCol))))  // so this line is only written once
```

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -144,7 +156,12 @@ class KMeansModel private[ml] (
+      new ArrayType(FloatType, true),
+      new ArrayType(FloatType, false))
+    SchemaUtils.checkColumnTypes(dataset.schema, $(featuresCol), typeCandidates)
     val data: RDD[OldVector] = dataset.select(col($(featuresCol))).rdd.map {
```
this won't take non-Vector types though; a unit test would catch this

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -312,6 +329,8 @@ class KMeans @Since("1.5.0") (
     val instances: RDD[OldVector] = dataset.select(col($(featuresCol))).rdd.map {
       case Row(point: Vector) => OldVectors.fromML(point)
+      case Row(point: Seq[_]) =>
+        OldVectors.fromML(Vectors.dense(point.asInstanceOf[Seq[Double]].toArray))
```
I'm not sure this will work with arrays of FloatType.  Make sure to test it

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -123,8 +128,15 @@ class KMeansModel private[ml] (
+      dataset.withColumn($(predictionCol), predictUDF(col($(featuresCol))))
+    } else {
+      val predictUDF = udf((vector: Seq[_]) =>
+        predict(Vectors.dense(vector.asInstanceOf[Seq[Double]].toArray)))
```
This may not work with arrays of FloatType.

jkbradley reviewed (commented):
Thanks for the PR!  I made a first review pass

SparkQA commented:
**[Test build #89410 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89410/testReport)** for PR 21081 at commit [`badb0cc`](https://github.com/apache/spark/commit/badb0cc5ca6ca69bb8e8fc0fce5ea05a4100bca0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89410/
Test PASSed.

lu-wang-dl committed:
make sure the code works for Float type and add the unit test

SparkQA commented:
**[Test build #89417 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89417/testReport)** for PR 21081 at commit [`6d222a3`](https://github.com/apache/spark/commit/6d222a3f257c850e653c9c048fb8c15e44d2c48f).

SparkQA commented:
**[Test build #89417 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89417/testReport)** for PR 21081 at commit [`6d222a3`](https://github.com/apache/spark/commit/6d222a3f257c850e653c9c048fb8c15e44d2c48f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -123,7 +128,21 @@ class KMeansModel private[ml] (
+    // val predictUDF = udf((vector: Vector) => predict(vector))
+    val predictUDF = if (dataset.schema($(featuresCol)).dataType.equals(new VectorUDT)) {
+      udf((vector: Vector) => predict(vector))
+    }
```
Scala style: } else {

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -123,7 +128,21 @@ class KMeansModel private[ml] (
+      udf((vector: Vector) => predict(vector))
+    }
+    else {
+      udf((vector: Seq[_]) => {
```
scala style: remove unnecessary ```{``` at end of line (IntelliJ should warn you about this)

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -90,7 +90,12 @@ private[clustering] trait KMeansParams extends Params with HasMaxIter with HasFe
   protected def validateAndTransformSchema(schema: StructType): StructType = {
-    SchemaUtils.checkColumnType(schema, $(featuresCol), new VectorUDT)
+    val typeCandidates = List( new VectorUDT,
+      new ArrayType(DoubleType, true),
```
Thinking about this, let's actually disallow nullable columns.  KMeans won't handle nulls properly.

jkbradley replied:
Also, IntelliJ may warn you about passing boolean arguments as named arguments; that'd be nice to fix here.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -123,7 +128,21 @@ class KMeansModel private[ml] (
+    }
+    else {
+      udf((vector: Seq[_]) => {
+        val featureArray = Array.fill[Double](vector.size)(0.0)
```
You shouldn't have to do the conversion in this convoluted (and less efficient) way.  I'd recommend doing a match-case statement on dataset.schema; I think that will be the most succinct.  Then you can handle Vector, Seq of Float, and Seq of Double separately, without conversions to strings.

Same for the similar cases below.

jkbradley reviewed (commented):
Thanks for the updates!

jkbradley commented:
I hope we can apply it to other algs too.  @ludatabricks is doing some refactoring which should make that easier, but we're not going for a completely general approach right away.

I don't think we need to worry about sparse FloatType features; users have no way to pass those in.

WeichenXu123 commented:
So why not design generic vector class ? and then implement Vector[Double] and Vector[Float] via generic specification ? So it can support everything, no matter sparse and dense.

lu-wang-dl committed:
consolidating featuretovector

SparkQA commented:
**[Test build #89585 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89585/testReport)** for PR 21081 at commit [`009b918`](https://github.com/apache/spark/commit/009b918c8734b19f9f9b34a31c23d6ad582c7465).

SparkQA commented:
**[Test build #89585 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89585/testReport)** for PR 21081 at commit [`009b918`](https://github.com/apache/spark/commit/009b918c8734b19f9f9b34a31c23d6ad582c7465).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89585/
Test PASSed.

jkbradley commented:
@WeichenXu123  A generic vector class would be interesting, but that would be a big project, way out of scope of this PR.  You could bring it up if that person on the dev list sends a SPIP about linear algebra.

jkbradley reviewed (commented):
A few more comments.  (Also, remember to clean up the commented-out code.)

lu-wang-dl committed:
change featureToVector to KMeanParams and add the scala docs
add validateSchema and use it in computeCost

addressed the comments from @jkbradley

SparkQA commented:
**[Test build #89660 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89660/testReport)** for PR 21081 at commit [`cd988c7`](https://github.com/apache/spark/commit/cd988c7f2e4c2cb1f2006e264a24529a72c9a5cf).

SparkQA commented:
**[Test build #89660 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89660/testReport)** for PR 21081 at commit [`cd988c7`](https://github.com/apache/spark/commit/cd988c7f2e4c2cb1f2006e264a24529a72c9a5cf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89660/
Test PASSed.

lu-wang-dl committed:
change featureToVector and validateSchema to private functions

SparkQA commented:
**[Test build #89736 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89736/testReport)** for PR 21081 at commit [`3ffb322`](https://github.com/apache/spark/commit/3ffb32291503a628c63a7014d27f2313f04c5497).

lu-wang-dl committed:
Merge branch 'master' of https://github.com/apache/spark into SPARK-23975

SparkQA commented:
**[Test build #89738 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89738/testReport)** for PR 21081 at commit [`fee36ad`](https://github.com/apache/spark/commit/fee36ad2deaa013c8fa61cae399a274cd7681fa9).

lu-wang-dl committed:
move featureToVector to util, so that other methods could use it to allow array input

SparkQA commented:
**[Test build #89743 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89743/testReport)** for PR 21081 at commit [`3e012fb`](https://github.com/apache/spark/commit/3e012fba3470c0f938e33cd3c783dff5ee068fcf).

SparkQA commented:
**[Test build #89738 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89738/testReport)** for PR 21081 at commit [`fee36ad`](https://github.com/apache/spark/commit/fee36ad2deaa013c8fa61cae399a274cd7681fa9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89738/
Test PASSed.

SparkQA commented:
**[Test build #89743 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89743/testReport)** for PR 21081 at commit [`3e012fb`](https://github.com/apache/spark/commit/3e012fba3470c0f938e33cd3c783dff5ee068fcf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89743/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -86,13 +86,23 @@ private[clustering] trait KMeansParams extends Params with HasMaxIter with HasFe
+      new ArrayType(FloatType, false))
+    SchemaUtils.checkColumnTypes(schema, $(featuresCol), typeCandidates)
+  }
   /**
```
scala style: always put newline between methods

jkbradley replied:
Ping: There needs to be a newline between the "}" of the previous method and the "/**" Scaladoc of the next method.  Please start checking for this.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/util/DatasetUtils.scala:
```diff
@@ -0,0 +1,54 @@
+   * @param colName column name for features
+   * @return Vector feature column
+   */
+  @Since("2.4.0")
```
Don't add Since annotations to private APIs.  They can get Since annotations when they are made public.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/util/DatasetUtils.scala:
```diff
@@ -0,0 +1,54 @@
+private[spark] object DatasetUtils {
+
+  /**
+   * preprocessing the input feature column to Vector
```
This is a bit unclear.  How about: "Cast a column in a Dataset to a Vector type."
Also, this isn't specific to features, so please clarify that below.
Finally, the key thing to document is the list of supported input types, so I'd add that.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/util/DatasetUtils.scala:
```diff
@@ -0,0 +1,54 @@
+   * preprocessing the input feature column to Vector
+   * @param dataset DataFrame with columns for features
+   * @param colName column name for features
+   * @return Vector feature column
```
Add a note that this returned Column does not have Metadata

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/clustering/KMeansSuite.scala:
```diff
@@ -199,6 +201,47 @@ class KMeansSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultR
+    assert(newdatasetD.schema(featuresColNameD).dataType.equals(new ArrayType(DoubleType, false)))
+    assert(newdatasetF.schema(featuresColNameF).dataType.equals(new ArrayType(FloatType, false)))
+
+    val kmeansD = new KMeans().setK(k).setFeaturesCol(featuresColNameD).setSeed(1)
```
Also do: `setMaxIter(1)` to make this a little faster.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/clustering/KMeansSuite.scala:
```diff
@@ -199,6 +201,47 @@ class KMeansSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultR
+    assert(predictDifference.count() == 0)
+
+    assert(modelD.computeCost(newdatasetD) == modelF.computeCost(newdatasetF) )
+
```
nit: remove unnecessary newline

jkbradley reviewed (commented):
Thanks for the updates!  After these small fixes, this should be ready, and then we can continue with other algorithms.

lu-wang-dl committed:
fix the comments and correct code style

SparkQA commented:
**[Test build #89749 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89749/testReport)** for PR 21081 at commit [`c4e1a51`](https://github.com/apache/spark/commit/c4e1a51551993008d7b082b112d2296cbc4eb97b).

SparkQA commented:
**[Test build #89736 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89736/testReport)** for PR 21081 at commit [`3ffb322`](https://github.com/apache/spark/commit/3ffb32291503a628c63a7014d27f2313f04c5497).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89736/
Test FAILed.

SparkQA commented:
**[Test build #89749 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89749/testReport)** for PR 21081 at commit [`c4e1a51`](https://github.com/apache/spark/commit/c4e1a51551993008d7b082b112d2296cbc4eb97b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89749/
Test PASSed.

SparkQA commented:
**[Test build #4157 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4157/testReport)** for PR 21081 at commit [`c4e1a51`](https://github.com/apache/spark/commit/c4e1a51551993008d7b082b112d2296cbc4eb97b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/util/DatasetUtils.scala:
```diff
@@ -27,28 +26,38 @@ import org.apache.spark.sql.types.{ArrayType, DoubleType, FloatType}
           case _: DoubleType => udf((vector: Seq[Double]) => {
             Vectors.dense(vector.toArray)
           })
+          case other =>
```
Thanks!  I forgot about this since this was generalized.

jkbradley reviewed (commented):
Just the 1 style comment.  Can you please fix it in the follow-up PR?  I'll go ahead and merge this with master
LGTM
Thanks @ludatabricks !
